### ISSUE TITLE
UPnP in the TcpConfig

### ISSUE THREAD

**Issue by tomaka** [libp2p/rust-libp2p#558]
Add `attempt_upnp: bool` in the `TcpConfig`, and implement it!

**Comment by tomaka** [comment]
Some references:
- http://www.upnp.org/specs/gw/UPnP-gw-WANIPConnection-v2-Service.pdf
- http://www.upnp-hacks.org/igd.html
- https://github.com/sbstp/rust-igd

One issue is that `listen_on` is only allowed to produce one listened address, while in practice if we traverse a NAT we are actually listening on two addresses: the one on the LAN and the on on the WAN.
Maybe `listen_on` should return an `Iterator<Item = Multiaddr>` instead of a `Multiaddr`.

**Comment by twittner** [comment]
How about [PCP](https://en.wikipedia.org/wiki/Port_Control_Protocol) instead of IGD?

**Comment by tomaka** [comment]
Whatever works!

**Comment by dvdplm** [comment]
@twittner I couldn't find any rust lib that implements PCP – do you know of any?

**Comment by Isan-Rivkin** [comment]
@dvdplm 
*Maybe* this one [pcp_mmv-rs](https://github.com/saurvs/pcp_mmv-rs)
Or you could maybe use FFI 
- [pcp](https://github.com/libpcp/pcp)
- [repcpc](https://github.com/alfredh/repcpc)

**Comment by twittner** [comment]
@dvdplm: No, I do not. FYI, I have started a small UPNP-IGD implementation at https://github.com/paritytech/upnp-igdp but it is currently using a fork of tokio and I am reluctant to test it, given the security implications of opening port mappings available to the outside world.

**Comment by dvdplm** [comment]
> fork of tokio

Curious: why did you have to fork it?

**Comment by tomaka** [comment]
> Curious: why did you have to fork it?

https://github.com/tokio-rs/tokio/pull/710

**Comment by tomaka** [comment]
I'm kind of conflicted whether this is a good "getting-started" issue.

The problem we encountered is that opening on the router is not instantaneous.
When one calls `listen_on` on `TcpConfig`, it needs to open ports. However the `listen_on` method is expected to immediately return the listened address, which depends on opening ports.

This also raises the question of making `listen_on` report known external addresses, for example the IP addresses of the local network interfaces.

**Comment by thomaseizinger** [comment]
Closed as stale.
